# TCC - Genetic Algorithm to control Waves in Games

### Authors:
 - Daniel Hotta
 - Rafael Gonçalves Pereira Silva
 - Ricardo Akira Tanaka

In [71]:
import os
import numpy as np
from numpy.polynomial import Polynomial
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import matplotlib.gridspec as mpgs
from matplotlib.backends.backend_pdf import PdfPages


## Simple description

This notebook uses the results of the experiments we made with our Genetic Algorithm to conclude the overall perfomance agrandomnst a Uniform Enemy Generator and "Naive Strategies" (Like All Enemies with the same color or One of Each color).
For uniform enemies, 10 waves were enough data, otherwise (random and AI) 30 waves were collected.

- The libs you can find in two repositories: [SpaceShip](https://github.com/RGPRafael/godot) and [Tower Defense](https://github.com/raktanaka/tccTD)

- The full dataset can be found [here](https://github.com/raktanaka/tcc-results).




Reads txt data, trimmings results for 300 lines in case of 10 waves, or 900 lines for 30 waves.

In [72]:
def GetFileData(filepath):
    # print(filepath)
    df = pd.read_csv(filepath, header=None, delimiter=';')

    if df.shape[0] < 899:
        return(df.iloc[:300])
    else:
        return(df.iloc[:900])


Getting files for each version of fitness tested:
- 1, 2, 3 for TD
- 1, 2 fot SS

In [73]:
td_allred_rd_file = os.path.join(
    'Fitness Tests', 'td_rd', 'AllRed - Random.txt')
td_allgreen_rd_file = os.path.join(
    'Fitness Tests', 'td_rd', 'AllGreen - Random.txt')
td_greenred_rd_file = os.path.join(
    'Fitness Tests', 'td_rd', 'GreenRed - Random.txt')
td_redgreen_rd_file = os.path.join(
    'Fitness Tests', 'td_rd', 'RedGreen - Random.txt')

# ---------------------------------------------------------------------------- #

td_allred_ai_file_1 = os.path.join(
    'Fitness Tests', 'td_v1', 'AllRed - AI.txt')
td_allgreen_ai_file_1 = os.path.join(
    'Fitness Tests', 'td_v1', 'AllGreen - AI.txt')
td_greenred_ai_file_1 = os.path.join(
    'Fitness Tests', 'td_v1', 'GreenRed - AI.txt')
td_redgreen_ai_file_1 = os.path.join(
    'Fitness Tests', 'td_v1', 'RedGreen - AI.txt')

# ---------------------------------------------------------------------------- #

td_allred_ai_file_2 = os.path.join(
    'Fitness Tests', 'td_v2', 'AllRed - AI.txt')
td_allgreen_ai_file_2 = os.path.join(
    'Fitness Tests', 'td_v2', 'AllGreen - AI.txt')
td_greenred_ai_file_2 = os.path.join(
    'Fitness Tests', 'td_v2', 'GreenRed - AI.txt')
td_redgreen_ai_file_2 = os.path.join(
    'Fitness Tests', 'td_v2', 'RedGreen - AI.txt')

# ---------------------------------------------------------------------------- #

td_allred_ai_file_3 = os.path.join(
    'Fitness Tests', 'td_v3', 'AllRed - AI.txt')
td_allgreen_ai_file_3 = os.path.join(
    'Fitness Tests', 'td_v3', 'AllGreen - AI.txt')
td_greenred_ai_file_3 = os.path.join(
    'Fitness Tests', 'td_v3', 'GreenRed - AI.txt')
td_redgreen_ai_file_3 = os.path.join(
    'Fitness Tests', 'td_v3', 'RedGreen - AI.txt')


In [74]:
ss_yellowmove_rd_file = os.path.join(
    'Fitness Tests', 'ss_rd', 'IA_MOVING_ SHOOTING _YELLOW_SHOOT - Random.txt')
ss_yellowstill_rd_file = os.path.join(
    'Fitness Tests', 'ss_rd', 'IA_STILL_YELLOW_SHOOT - Random.txt')
ss_redmove_rd_file = os.path.join(
    'Fitness Tests', 'ss_rd', 'IA_MOVING_SHOOTING_ RED_SHOOT - Random.txt')
ss_redstill_rd_file = os.path.join(
    'Fitness Tests', 'ss_rd', 'IA_STILL _RED_SHOOT - Random.txt')

# ---------------------------------------------------------------------------- #

ss_yellowmove_ai_file_1 = os.path.join(
    'Fitness Tests', 'ss_v1', 'IA_MOVING_ SHOOTING _YELLOW_SHOOT - AI.txt')
ss_yellowstill_ai_file_1 = os.path.join(
    'Fitness Tests', 'ss_v1', 'IA_STILL_YELLOW_SHOOT - AI.txt')
ss_redmove_ai_file_1 = os.path.join(
    'Fitness Tests', 'ss_v1', 'IA_MOVING_SHOOTING_ RED_SHOOT - AI.txt')
ss_redstill_ai_file_1 = os.path.join(
    'Fitness Tests', 'ss_v1', 'IA_STILL _RED_SHOOT - AI.txt')

# ---------------------------------------------------------------------------- #

ss_yellowmove_ai_file_2 = os.path.join(
    'Fitness Tests', 'ss_v2', 'IA_MOVING_ SHOOTING _YELLOW_SHOOT - AI.txt')
ss_yellowstill_ai_file_2 = os.path.join(
    'Fitness Tests', 'ss_v2', 'IA_STILL_YELLOW_SHOOT - AI.txt')
ss_redmove_ai_file_2 = os.path.join(
    'Fitness Tests', 'ss_v2', 'IA_MOVING_SHOOTING_ RED_SHOOT - AI.txt')
ss_redstill_ai_file_2 = os.path.join(
    'Fitness Tests', 'ss_v2', 'IA_STILL _RED_SHOOT - AI.txt')


Tower Defense data "header"
0: wave number
1-12: element in wave
13: damage done in wave

In [75]:
td_allred_rd = GetFileData(td_allred_rd_file)
td_allgreen_rd = GetFileData(td_allgreen_rd_file)
td_greenred_rd = GetFileData(td_greenred_rd_file)
td_redgreen_rd = GetFileData(td_redgreen_rd_file)

# ---------------------------------------------------------------------------- #

td_allred_ai_1 = GetFileData(td_allred_ai_file_1)
td_allgreen_ai_1 = GetFileData(td_allgreen_ai_file_1)
td_greenred_ai_1 = GetFileData(td_greenred_ai_file_1)
td_redgreen_ai_1 = GetFileData(td_redgreen_ai_file_1)

# ---------------------------------------------------------------------------- #

td_allred_ai_2 = GetFileData(td_allred_ai_file_2)
td_allgreen_ai_2 = GetFileData(td_allgreen_ai_file_2)
td_greenred_ai_2 = GetFileData(td_greenred_ai_file_2)
td_redgreen_ai_2 = GetFileData(td_redgreen_ai_file_2)

# ---------------------------------------------------------------------------- #

td_allred_ai_3 = GetFileData(td_allred_ai_file_3)
td_allgreen_ai_3 = GetFileData(td_allgreen_ai_file_3)
td_greenred_ai_3 = GetFileData(td_greenred_ai_file_3)
td_redgreen_ai_3 = GetFileData(td_redgreen_ai_file_3)


In [76]:
ss_yellowmove_rd = GetFileData(ss_yellowmove_rd_file)
ss_yellowstill_rd = GetFileData(ss_yellowstill_rd_file)
ss_redmove_rd = GetFileData(ss_redmove_rd_file)
ss_redstill_rd = GetFileData(ss_redstill_rd_file)

# ---------------------------------------------------------------------------- #

ss_yellowmove_ai_1 = GetFileData(ss_yellowmove_ai_file_1)
ss_yellowstill_ai_1 = GetFileData(ss_yellowstill_ai_file_1)
ss_redmove_ai_1 = GetFileData(ss_redmove_ai_file_1)
ss_redstill_ai_1 = GetFileData(ss_redstill_ai_file_1)

# ---------------------------------------------------------------------------- #

ss_yellowmove_ai_2 = GetFileData(ss_yellowmove_ai_file_2)
ss_yellowstill_ai_2 = GetFileData(ss_yellowstill_ai_file_2)
ss_redmove_ai_2 = GetFileData(ss_redmove_ai_file_2)
ss_redstill_ai_2 = GetFileData(ss_redstill_ai_file_2)


In [77]:
def GenerateHeader(df):
    header = ['wave number']

    if df.shape[1] < 14:
        n_enemies = 6
    else:
        n_enemies = 12

    for each in range(n_enemies):
        header.append(str('enemy ' + str(each + 1)))

    header.append('total damage')
    return(header)


In [78]:
td_allred_rd.columns = GenerateHeader(td_allred_rd)
td_allgreen_rd.columns = GenerateHeader(td_allgreen_rd)
td_greenred_rd.columns = GenerateHeader(td_greenred_rd)
td_redgreen_rd.columns = GenerateHeader(td_redgreen_rd)

# ---------------------------------------------------------------------------- #

td_allred_ai_1.columns = GenerateHeader(td_allred_ai_1)
td_allgreen_ai_1.columns = GenerateHeader(td_allgreen_ai_1)
td_greenred_ai_1.columns = GenerateHeader(td_greenred_ai_1)
td_redgreen_ai_1.columns = GenerateHeader(td_redgreen_ai_1)

# ---------------------------------------------------------------------------- #

td_allred_ai_2.columns = GenerateHeader(td_allred_ai_2)
td_allgreen_ai_2.columns = GenerateHeader(td_allgreen_ai_2)
td_greenred_ai_2.columns = GenerateHeader(td_greenred_ai_2)
td_redgreen_ai_2.columns = GenerateHeader(td_redgreen_ai_2)

# ---------------------------------------------------------------------------- #

td_allred_ai_3.columns = GenerateHeader(td_allred_ai_3)
td_allgreen_ai_3.columns = GenerateHeader(td_allgreen_ai_3)
td_greenred_ai_3.columns = GenerateHeader(td_greenred_ai_3)
td_redgreen_ai_3.columns = GenerateHeader(td_redgreen_ai_3)


In [79]:
ss_yellowmove_rd.columns = GenerateHeader(ss_yellowmove_rd)
ss_yellowstill_rd.columns = GenerateHeader(ss_yellowstill_rd)
ss_redmove_rd.columns = GenerateHeader(ss_redmove_rd)
ss_redstill_rd.columns = GenerateHeader(ss_redstill_rd)

# ---------------------------------------------------------------------------- #

ss_yellowmove_ai_1.columns = GenerateHeader(ss_yellowmove_ai_1)
ss_yellowstill_ai_1.columns = GenerateHeader(ss_yellowstill_ai_1)
ss_redmove_ai_1.columns = GenerateHeader(ss_redmove_ai_1)
ss_redstill_ai_1.columns = GenerateHeader(ss_redstill_ai_1)

# ---------------------------------------------------------------------------- #

ss_yellowmove_ai_2.columns = GenerateHeader(ss_yellowmove_ai_2)
ss_yellowstill_ai_2.columns = GenerateHeader(ss_yellowstill_ai_2)
ss_redmove_ai_2.columns = GenerateHeader(ss_redmove_ai_2)
ss_redstill_ai_2.columns = GenerateHeader(ss_redstill_ai_2)


Calculates statistics from the i-wave from each of the 30 experiments

In [80]:
def CalcStats(df):

    new_df_header = ['wave number', 'average damage', 'standart deviation']
    wave_num = list(range(1, 31))
    mean = []
    stdev = []

    # Loops each i wave for the 30 experiments
    for i in range(1, 31):
        tmp_df = df[df['wave number'] == i]
        mean.append(tmp_df['total damage'].mean())
        stdev.append(tmp_df['total damage'].std())
        #l['average damage'] = f"{tmp_df['total damage'].mean():.2f}"
        #l['standart deviation'] = f"{tmp_df['total damage'].std():.2f}"
        # res.append(l)

    new_df = pd.DataFrame(zip(wave_num, mean, stdev), columns=new_df_header)
    return(new_df)


In [81]:
td_allgreen_rd_stats = CalcStats(td_allgreen_rd)
td_allred_rd_stats = CalcStats(td_allred_rd)
td_greenred_rd_stats = CalcStats(td_greenred_rd)
td_redgreen_rd_stats = CalcStats(td_redgreen_rd)

# ---------------------------------------------------------------------------- #

td_allgreen_ai_stats_1 = CalcStats(td_allgreen_ai_1)
td_allred_ai_stats_1 = CalcStats(td_allred_ai_1)
td_greenred_ai_stats_1 = CalcStats(td_greenred_ai_1)
td_redgreen_ai_stats_1 = CalcStats(td_redgreen_ai_1)

# ---------------------------------------------------------------------------- #

td_allgreen_ai_stats_2 = CalcStats(td_allgreen_ai_2)
td_allred_ai_stats_2 = CalcStats(td_allred_ai_2)
td_greenred_ai_stats_2 = CalcStats(td_greenred_ai_2)
td_redgreen_ai_stats_2 = CalcStats(td_redgreen_ai_2)

# ---------------------------------------------------------------------------- #

td_allgreen_ai_stats_3 = CalcStats(td_allgreen_ai_3)
td_allred_ai_stats_3 = CalcStats(td_allred_ai_3)
td_greenred_ai_stats_3 = CalcStats(td_greenred_ai_3)
td_redgreen_ai_stats_3 = CalcStats(td_redgreen_ai_3)


In [82]:
ss_yellowmove_rd_stats = CalcStats(ss_yellowmove_rd)
ss_yellowstill_rd_stats = CalcStats(ss_yellowstill_rd)
ss_redmove_rd_stats = CalcStats(ss_redmove_rd)
ss_redstill_rd_stats = CalcStats(ss_redstill_rd)

# ---------------------------------------------------------------------------- #

ss_yellowmove_ai_stats_1 = CalcStats(ss_yellowmove_ai_1)
ss_yellowstill_ai_stats_1 = CalcStats(ss_yellowstill_ai_1)
ss_redmove_ai_stats_1 = CalcStats(ss_redmove_ai_1)
ss_redstill_ai_stats_1 = CalcStats(ss_redstill_ai_1)

# ---------------------------------------------------------------------------- #

ss_yellowmove_ai_stats_2 = CalcStats(ss_yellowmove_ai_2)
ss_yellowstill_ai_stats_2 = CalcStats(ss_yellowstill_ai_2)
ss_redmove_ai_stats_2 = CalcStats(ss_redmove_ai_2)
ss_redstill_ai_stats_2 = CalcStats(ss_redstill_ai_2)


In [83]:
def GetDataPerWave(df):

    dic = {}

    # Loops each i wave for the 30 experiments
    for i in range(1, 31):
        tmp_df = df[df['wave number'] == i]
        dic[str(i)] = tmp_df['total damage'].values
        #l.append(tmp_df['total damage'].values)

    return(pd.DataFrame.from_dict(dic))


In [84]:
td_allgreen_rd_perwave = GetDataPerWave(td_allgreen_rd)
td_allred_rd_perwave = GetDataPerWave(td_allred_rd)
td_greenred_rd_perwave = GetDataPerWave(td_greenred_rd)
td_redgreen_rd_perwave = GetDataPerWave(td_redgreen_rd)

# ---------------------------------------------------------------------------- #

td_allgreen_ai_perwave_1 = GetDataPerWave(td_allgreen_ai_1)
td_allred_ai_perwave_1 = GetDataPerWave(td_allred_ai_1)
td_greenred_ai_perwave_1 = GetDataPerWave(td_greenred_ai_1)
td_redgreen_ai_perwave_1 = GetDataPerWave(td_redgreen_ai_1)

# ---------------------------------------------------------------------------- #

td_allgreen_ai_perwave_2 = GetDataPerWave(td_allgreen_ai_2)
td_allred_ai_perwave_2 = GetDataPerWave(td_allred_ai_2)
td_greenred_ai_perwave_2 = GetDataPerWave(td_greenred_ai_2)
td_redgreen_ai_perwave_2 = GetDataPerWave(td_redgreen_ai_2)

# ---------------------------------------------------------------------------- #

td_allgreen_ai_perwave_3 = GetDataPerWave(td_allgreen_ai_3)
td_allred_ai_perwave_3 = GetDataPerWave(td_allred_ai_3)
td_greenred_ai_perwave_3 = GetDataPerWave(td_greenred_ai_3)
td_redgreen_ai_perwave_3 = GetDataPerWave(td_redgreen_ai_3)


In [85]:
ss_yellowmove_rd_perwave = GetDataPerWave(ss_yellowmove_rd)
ss_yellowstill_rd_perwave = GetDataPerWave(ss_yellowstill_rd)
ss_redmove_rd_perwave = GetDataPerWave(ss_redmove_rd)
ss_redstill_rd_perwave = GetDataPerWave(ss_redstill_rd)

# ---------------------------------------------------------------------------- #

ss_yellowmove_ai_perwave_1 = GetDataPerWave(ss_yellowmove_ai_1)
ss_yellowstill_ai_perwave_1 = GetDataPerWave(ss_yellowstill_ai_1)
ss_redmove_ai_perwave_1 = GetDataPerWave(ss_redmove_ai_1)
ss_redstill_ai_perwave_1 = GetDataPerWave(ss_redstill_ai_1)

# ---------------------------------------------------------------------------- #

ss_yellowmove_ai_perwave_2 = GetDataPerWave(ss_yellowmove_ai_2)
ss_yellowstill_ai_perwave_2 = GetDataPerWave(ss_yellowstill_ai_2)
ss_redmove_ai_perwave_2 = GetDataPerWave(ss_redmove_ai_2)
ss_redstill_ai_perwave_2 = GetDataPerWave(ss_redstill_ai_2)


In [86]:
def GetStdMax(df):
    return(np.ceil(float(df['standart deviation'].max())))


def GetAvgMin(df):
    return(np.floor(float(df['average damage'].min())))


def GetAvgMax(df):
    return(np.ceil(float(df['average damage'].max())))


def GetAxesValues(df_rd, df_1, df_2, df_3=pd.DataFrame()):

    std_rd = GetStdMax(df_rd)
    std_1 = GetStdMax(df_1)
    std_2 = GetStdMax(df_2)
    min_rd = GetAvgMin(df_rd)
    min_1 = GetAvgMin(df_1)
    min_2 = GetAvgMin(df_2)
    max_rd = GetAvgMax(df_rd)
    max_1 = GetAvgMax(df_1)
    max_2 = GetAvgMax(df_2)

    if not df_3.empty:
        std_3 = GetStdMax(df_3)
        min_3 = GetAvgMin(df_3)
        max_3 = GetAvgMax(df_3)
        std_max = max([std_rd, std_1, std_2, std_3])
        dmg_min = min([min_rd, min_1, min_2, min_3])
        dmg_max = max([max_rd, max_1, max_2, max_3])
    else:
        std_max = max([std_rd, std_1, std_2])
        dmg_min = min([min_rd, min_1, min_2])
        dmg_max = max([max_rd, max_1, max_2])

    return(dmg_min, dmg_max, std_max)



In [100]:
def PlotFitnessTest(title, df_rd, rd_max, df_1, df_2, df_3=pd.DataFrame()):

    fig, ax = plt.subplots(figsize=(15, 8), dpi=80)
    ax.set_title(title, fontdict=None, loc='center', pad=None)

    color1 = 'royalblue'
    color2 = 'green'
    color3 = 'darkorange'

    rd_line_max = 30 * [rd_max]
    rd_avg = float(df_rd['average damage'].mean())
    rd_line_avg = 30 * [rd_avg]

    ax.plot(df_rd['wave number'], rd_line_max, marker='None',
             linestyle='dotted', color='grey', label='Máximo no Aleatório')
    ax.plot(df_rd['wave number'], rd_line_avg, marker='None',
             linestyle='solid', color='grey', label='Média do Aleatório')

    ax.plot(df_1['wave number'], df_1['average damage'],
             marker='o', linestyle='None', color=color1, label='v1')
    fit = Polynomial.fit(df_1['wave number'], df_1['average damage'], deg=5)
    ax.plot(df_1['wave number'], fit(df_1['wave number']), color=color1)

    if not df_3.empty:
        ax.plot(df_2['wave number'], df_2['average damage'],
             marker='s', linestyle='None', color=color2, label='v2')
        fit = Polynomial.fit(df_2['wave number'], df_2['average damage'], deg=5)
        ax.plot(df_2['wave number'], fit(df_2['wave number']), color=color2)

        ax.plot(df_3['wave number'], df_3['average damage'], 
             marker='D', linestyle='None', color=color3, label='v3')
        fit = Polynomial.fit(df_3['wave number'], df_3['average damage'], deg=5)
        ax.plot(df_3['wave number'], fit(df_3['wave number']), color=color3)
    else:
        ax.plot(df_2['wave number'], df_2['average damage'],
             marker='D', linestyle='None', color=color3, label='v3')
        fit = Polynomial.fit(df_2['wave number'], df_2['average damage'], deg=5)
        ax.plot(df_2['wave number'], fit(df_2['wave number']), color=color3)

    dmg_min, dmg_max, std_dev = GetAxesValues(df_rd, df_1, df_2, df_3)
    ax.set_xlabel('Wave Number')
    ax.set_xticks(df_rd['wave number'])
    ax.set_ylabel('Average Damage')
    ax.set_yticks(np.arange(np.floor(0.8 * min(dmg_min, rd_avg)), np.ceil(1.2 * max(dmg_max, rd_max)), (dmg_max - dmg_min) // 10))

    ax.legend(loc='upper left', borderaxespad=0.5)
    #plt.show()
    plt.savefig(title + ' Fitness', facecolor='white')
    plt.close()

In [88]:
td_allgreen_rd_max = td_allgreen_rd['total damage'].max()
td_allred_rd_max = td_allred_rd['total damage'].max()
td_greenred_rd_max = td_greenred_rd['total damage'].max()
td_redgreen_rd_max = td_redgreen_rd['total damage'].max()

ss_yellowmove_rd_max = ss_yellowmove_rd['total damage'].max()
ss_yellowstill_rd_max = ss_yellowstill_rd['total damage'].max()
ss_redmove_rd_max = ss_redmove_rd['total damage'].max()
ss_redstill_rd_max = ss_redstill_rd['total damage'].max()

In [101]:
PlotFitnessTest('Torres Verdes', td_allgreen_rd_stats, float(td_allgreen_rd_max), td_allgreen_ai_stats_1, td_allgreen_ai_stats_2, td_allgreen_ai_stats_3)
PlotFitnessTest('Torres Vermelhas', td_allred_rd_stats, float(td_allred_rd_max), td_allred_ai_stats_1, td_allred_ai_stats_2, td_allred_ai_stats_3)
PlotFitnessTest('Torres Verdes e Vermelhas', td_greenred_rd_stats, float(td_greenred_rd_max), td_greenred_ai_stats_1, td_greenred_ai_stats_2, td_greenred_ai_stats_3)
PlotFitnessTest('Torres Vermelhas e Verdes', td_redgreen_rd_stats, float(td_redgreen_rd_max), td_redgreen_ai_stats_1, td_redgreen_ai_stats_2, td_redgreen_ai_stats_3)


In [102]:
PlotFitnessTest('Nave com Disparo Amarelo e Movendo', ss_yellowmove_rd_stats, float(ss_yellowmove_rd_max), ss_yellowmove_ai_stats_1, ss_yellowmove_ai_stats_2)
PlotFitnessTest('Nave com Disparo Amarelo e Parada', ss_yellowstill_rd_stats, float(ss_yellowstill_rd_max), ss_yellowstill_ai_stats_1, ss_yellowstill_ai_stats_2)
PlotFitnessTest('Nave com Disparo Vermelho e Movendo', ss_redmove_rd_stats, float(ss_redmove_rd_max), ss_redmove_ai_stats_1, ss_redmove_ai_stats_2)
PlotFitnessTest('Nave com Disparo Vermelho e Parada', ss_redstill_rd_stats, float(ss_redstill_rd_max), ss_redstill_ai_stats_1, ss_redstill_ai_stats_2)

In [97]:
display(td_allgreen_rd[td_allgreen_rd['total damage'] == td_allgreen_rd_max])
display(td_allred_rd[td_allred_rd['total damage'] == td_allred_rd_max])
display(td_greenred_rd[td_greenred_rd['total damage'] == td_greenred_rd_max])
display(td_redgreen_rd[td_redgreen_rd['total damage'] == td_redgreen_rd_max])

wave number           enemy 1           enemy 2           enemy 3  \
135           16   [EnemyGreen, 1]   [EnemyGreen, 1]   [EnemyGreen, 0]   

            enemy 4            enemy 5           enemy 6         enemy 7  \
135   [EnemyRed, 1]   [EnemyPurple, 0]   [EnemyGreen, 0]   [EnemyRed, 1]   

              enemy 8            enemy 9          enemy 10          enemy 11  \
135   [EnemyGreen, 0]   [EnemyOrange, 0]   [EnemyGreen, 0]   [EnemyGreen, 0]   

           enemy 12  total damage  
135   [EnemyRed, 0]           290

wave number          enemy 1         enemy 2          enemy 3  \
618           19   [EnemyBlue, 0]   [EnemyRed, 0]   [EnemyBlue, 0]   

              enemy 4           enemy 5         enemy 6          enemy 7  \
618   [EnemyGreen, 0]   [EnemyGreen, 1]   [EnemyRed, 1]   [EnemyBlue, 0]   

               enemy 8          enemy 9           enemy 10           enemy 11  \
618   [EnemyOrange, 1]   [EnemyBlue, 0]   [EnemyYellow, 0]   [EnemyOrange, 0]   

              enemy 12  total damage  
618   [EnemyYellow, 1]           355

wave number            enemy 1          enemy 2          enemy 3  \
72           13   [EnemyPurple, 0]   [EnemyBlue, 0]   [EnemyBlue, 0]   

            enemy 4         enemy 5           enemy 6            enemy 7  \
72   [EnemyBlue, 1]   [EnemyRed, 0]   [EnemyGreen, 1]   [EnemyYellow, 0]   

           enemy 8           enemy 9          enemy 10          enemy 11  \
72   [EnemyRed, 1]   [EnemyGreen, 1]   [EnemyGreen, 0]   [EnemyGreen, 0]   

             enemy 12  total damage  
72   [EnemyYellow, 1]           330

wave number            enemy 1          enemy 2         enemy 3  \
577            8   [EnemyYellow, 0]   [EnemyBlue, 0]   [EnemyRed, 1]   

              enemy 4           enemy 5         enemy 6           enemy 7  \
577   [EnemyGreen, 0]   [EnemyGreen, 1]   [EnemyRed, 1]   [EnemyGreen, 0]   

               enemy 8         enemy 9          enemy 10          enemy 11  \
577   [EnemyYellow, 0]   [EnemyRed, 1]   [EnemyGreen, 0]   [EnemyGreen, 0]   

              enemy 12  total damage  
577   [EnemyPurple, 1]           315

In [192]:
display(ss_yellowmove_rd[ss_yellowmove_rd['total damage'] == ss_yellowmove_rd_max])
display(ss_yellowstill_rd[ss_yellowstill_rd['total damage'] == ss_yellowstill_rd_max])
display(ss_redmove_rd[ss_redmove_rd['total damage'] == ss_redmove_rd_max])
display(ss_redstill_rd[ss_redstill_rd['total damage'] == ss_redstill_rd_max])

wave number                   enemy 1                   enemy 2  \
193           14   [inimigo1, (-100, 300)]   [inimigo3, (-100, 100)]   

                      enemy 3                   enemy 4  \
193   [inimigo1, (-100, 300)]   [inimigo1, (-100, 300)]   

                    enemy 5                   enemy 6  total damage  
193   [inimigos, (90, -50)]   [inimigo3, (-100, 100)]           275

wave number                   enemy 1                   enemy 2  \
854           15   [inimigo3, (-100, 100)]   [inimigo1, (-100, 300)]   

                      enemy 3                   enemy 4  \
854   [inimigo3, (-100, 100)]   [inimigo1, (-100, 300)]   

                      enemy 5                   enemy 6  total damage  
854   [inimigo1, (-100, 300)]   [inimigo3, (-100, 100)]           315

wave number                   enemy 1                   enemy 2  \
870            1   [inimigo3, (-100, 100)]   [inimigo3, (-100, 100)]   

                      enemy 3                  enemy 4  \
870   [inimigo5, (1350, 500)]   [inimigo4, (200, -50)]   

                      enemy 5                   enemy 6  total damage  
870   [inimigo3, (-100, 100)]   [inimigo2, (1350, 100)]           260

wave number                 enemy 1                 enemy 2  \
393            4   [inimigo3, (90, -50)]   [inimigo1, (90, -50)]   

                    enemy 3                 enemy 4                 enemy 5  \
393   [inimigo3, (90, -50)]   [inimigo1, (90, -50)]   [inimigo5, (90, -50)]   

                    enemy 6  total damage  
393   [inimigo1, (90, -50)]           350

\begin{table}[]
\begin{tabular}{l|llllllllllll|l}
Modo             & Inimigo 1            & Inimigo 2           & Inimigo 3           & Inimigo 4           & Inimigo 5            & Inimigo 6           & Inimigo 7            & Inimigo 8            & Inimigo 9            & Inimigo 10           & Inimigo 11           & Inimigo 12           & Dano Total \\ \hline
Verde            & {[}EnemyGreen, 1{]}  & {[}EnemyGreen, 1{]} & {[}EnemyGreen, 0{]} & {[}EnemyRed, 1{]}   & {[}EnemyPurple, 0{]} & {[}EnemyGreen, 0{]} & {[}EnemyRed, 1{]}    & {[}EnemyGreen, 0{]}  & {[}EnemyOrange, 0{]} & {[}EnemyGreen, 0{]}  & {[}EnemyGreen, 0{]}  & {[}EnemyRed, 0{]}    & 290        \\
Vermelho         & {[}EnemyBlue, 0{]}   & {[}EnemyRed, 0{]}   & {[}EnemyBlue, 0{]}  & {[}EnemyGreen, 0{]} & {[}EnemyGreen, 1{]}  & {[}EnemyRed, 1{]}   & {[}EnemyBlue, 0{]}   & {[}EnemyOrange, 1{]} & {[}EnemyBlue, 0{]}   & {[}EnemyYellow, 0{]} & {[}EnemyOrange, 0{]} & {[}EnemyYellow, 1{]} & 355        \\
Verde + Vermelho & {[}EnemyPurple, 0{]} & {[}EnemyBlue, 0{]}  & {[}EnemyBlue, 0{]}  & {[}EnemyBlue, 1{]}  & {[}EnemyRed, 0{]}    & {[}EnemyGreen, 1{]} & {[}EnemyYellow, 0{]} & {[}EnemyRed, 1{]}    & {[}EnemyGreen, 1{]}  & {[}EnemyGreen, 0{]}  & {[}EnemyGreen, 0{]}  & {[}EnemyYellow, 1{]} & 330        \\
Vermelho + Verde & {[}EnemyYellow, 0{]} & {[}EnemyBlue, 0{]}  & {[}EnemyRed, 1{]}   & {[}EnemyGreen, 0{]} & {[}EnemyGreen, 1{]}  & {[}EnemyRed, 1{]}   & {[}EnemyGreen, 0{]}  & {[}EnemyYellow, 0{]} & {[}EnemyRed, 1{]}    & {[}EnemyGreen, 0{]}  & {[}EnemyGreen, 0{]}  & {[}EnemyPurple, 1{]} & 315       
\end{tabular}
\end{table}

In [20]:
def GetEnemyImg(enemy):

    #img_td_blue = mpimg.imread(os.path.join('sprites', 'tank_blue.png'))
    #img_td_green = mpimg.imread(os.path.join('sprites', 'tank_green.png'))
    #img_td_orange = mpimg.imread(os.path.join('sprites', 'tank_orange.png'))
    #img_td_purple = mpimg.imread(os.path.join('sprites', 'tank_purple.png'))
    #img_td_red = mpimg.imread(os.path.join('sprites', 'tank_red.png'))
    #img_td_yellow = mpimg.imread(os.path.join('sprites', 'tank_yellow.png'))

    #img_ss_inimigos = mpimg.imread(os.path.join('sprites', 'inimigos.png'))
    #img_ss_inimigo1 = mpimg.imread(os.path.join('sprites', 'inimigo1.png'))
    #img_ss_inimigo2 = mpimg.imread(os.path.join('sprites', 'inimigo2.png'))
    #img_ss_inimigo3 = mpimg.imread(os.path.join('sprites', 'inimigo3.png'))
    #img_ss_inimigo4 = mpimg.imread(os.path.join('sprites', 'inimigo4.png'))
    #img_ss_inimigo5 = mpimg.imread(os.path.join('sprites', 'inimigo5.png'))

    if enemy == 'EnemyRed':
        img = mpimg.imread(os.path.join('sprites', 'tank_red.png'))
    elif enemy == 'EnemyGreen':
        img = mpimg.imread(os.path.join('sprites', 'tank_green.png'))
    elif enemy == 'EnemyBlue':
        img = mpimg.imread(os.path.join('sprites', 'tank_blue.png'))
    elif enemy == 'EnemyYellow':
        img = mpimg.imread(os.path.join('sprites', 'tank_yellow.png'))
    elif enemy == 'EnemyPurple':
        img = mpimg.imread(os.path.join('sprites', 'tank_purple.png'))
    elif enemy == 'EnemyOrange':
        img = mpimg.imread(os.path.join('sprites', 'tank_orange.png'))

    elif enemy == 'inimigos':
        img = mpimg.imread(os.path.join('sprites', 'inimigos.png'))
    elif enemy == 'inimigo1':
        img = mpimg.imread(os.path.join('sprites', 'inimigo1.png'))
    elif enemy == 'inimigo2':
        img = mpimg.imread(os.path.join('sprites', 'inimigo2.png'))
    elif enemy == 'inimigo3':
        img = mpimg.imread(os.path.join('sprites', 'inimigo3.png'))
    elif enemy == 'inimigo4':
        img = mpimg.imread(os.path.join('sprites', 'inimigo4.png'))
    elif enemy == 'inimigo5':
        img = mpimg.imread(os.path.join('sprites', 'inimigo5.png'))

    return(img)

def GetDirection(dir):

    # TD path orientation
    if dir == '0':
        char = 'N'
    elif dir == '1':
        char = 'S'
    # SS asteroid spawn location
    #var inimigos_each = [['inimigos',Vector2(90,-50)  ],
    #                     ['inimigo1',Vector2(-100,300)],
    #                     ['inimigo2',Vector2(1350,100)],
    #                     ['inimigo3',Vector2(-100,100)],
    #                     ['inimigo4',Vector2(200,-50) ],
    #                     ['inimigo5',Vector2(1350,500)]]
    elif dir == '90,-50':
        char = '1'
    elif dir == '200,-50':
        char = '2'
    elif dir == '1350,100':
        char = '3'
    elif dir == '1350,500':
        char = '4'
    elif dir == '-100,300':
        char = '5'
    elif dir == '-100,100':
        char = '6'
    
    else:
        print(dir)
        char = '-'

    return(char)

In [32]:
def SplitRow(str):
    str = str.split(',')
    # Checks for TD data
    if len(str) == 2:
        return(str[0], str[1])
    else:
        return(str[0], (str[1] + ',' + str[2]))


def RemoveChar(str):

    ch_rem = '()[] '
    for i in ch_rem:
        str = str.replace(i, '')
    
    return(str)

def ListEnemies(df_row):

    enemy_list = []
    direction_list = []
    
    for each in df_row:
        str = RemoveChar(each)
        enemy, direction = SplitRow(str)
        enemy_list.append(enemy)
        direction_list.append(direction)

    return(enemy_list, direction_list)


def ImageWave(df, filename):
    n_lin, n_col = (df.shape)
    n_col = 2 * n_col - 2

    # Checks for 12 (TD) or 6 (SS) enemies per wave
    if n_col < 24:
        # Space Shooter
        #img_dpi = 200
        bg_color = 'xkcd:dark grey'
        font_color = 'white'
    else:
        # Tower Defense
        #img_dpi = 200
        bg_color = 'xkcd:light grey'
        font_color = 'black'

    #heights = [1] * n_lin
    fig = plt.figure(dpi=300, tight_layout=True)
    fig.patch.set_facecolor(bg_color)
    fig.set_size_inches(4, 0.5, forward=True)
    ax = np.zeros(n_col, dtype=object)
    gs = fig.add_gridspec(n_lin, n_col)
    # Loops each i wave mode, generating the images
    for i in range(n_lin):
        df_row = df.iloc[i, 1:-1]
        enemy_list, direction_list = ListEnemies(df_row)
        for j in range(n_col):
            ax[j] = fig.add_subplot(gs[i, j])
            plt.axis('off')
            if j == 0:
                ax[j].text(0.00, 0.45, str(i + 1), fontsize='xx-small', fontweight='bold', color=font_color)
            elif j == (n_col - 1):
                ax[j].text(0.45, 0.45, str(df.iloc[i, -1]), fontsize='xx-small', fontweight='bold', color=font_color)
            else:
            #ax[j].set_facecolor('xkcd:grey')
                #plt.axis('off')
                if j % 2 == 1:
                    nj = int(np.floor((j - 1) / 2))
                    img = GetEnemyImg(enemy_list[nj])
                    ax[j].imshow(img)
                else:
                    nj = (j // 2) - 1
                    char = GetDirection(direction_list[nj])
                    ax[j].text(-0.05, 0.44, char, fontsize=4.0, fontweight='bold', color=font_color)
                    #tail, head = GetDirection(direction_list[nj])
                    #ax[j].add_patch(mpatches.FancyArrowPatch(tail, head, mutation_scale=4, color='black'))
    plt.close()
    filename += '.png'
    fig.savefig(filename)


In [33]:
ImageWave(td_allgreen_rd[td_allgreen_rd['total damage'] == td_allgreen_rd_max], 'td_allgreen_ai_large')
ImageWave(td_allred_rd[td_allred_rd['total damage'] == td_allred_rd_max], 'td_allred_ai_large')
ImageWave(td_greenred_rd[td_greenred_rd['total damage'] == td_greenred_rd_max], 'td_greenred_ai_large')
ImageWave(td_redgreen_rd[td_redgreen_rd['total damage'] == td_redgreen_rd_max], 'td_redgreen_ai_large')

ImageWave(ss_yellowmove_rd[ss_yellowmove_rd['total damage'] == ss_yellowmove_rd_max], 'ss_yellowmove_ai_large')
ImageWave(ss_yellowstill_rd[ss_yellowstill_rd['total damage'] == ss_yellowstill_rd_max], 'ss_yellowstill_ai_large')
ImageWave(ss_redmove_rd[ss_redmove_rd['total damage'] == ss_redmove_rd_max], 'ss_redmove_ai_large')
ImageWave(ss_redstill_rd[ss_redstill_rd['total damage'] == ss_redstill_rd_max], 'ss_redstill_ai_large')

/tmp/ipykernel_14735/3233563230.py:80: UserWarning: Tight layout not applied. tight_layout cannot make axes width small enough to accommodate all axes decorations
  fig.savefig(filename)


Visually the wave were stability was obtained wac calculated
|             | v1 | v2 | v3 |
|-------------|----|----|----|
| Green       | 8  | 12 | 8  |
| Red         | 15 | -  | 12 |
| Green + Red | 18 | -  | 11 |
| Red + Green | 21 | -  | 12 |

In [108]:
td_allgreen_ai_stats_1['average damage'][7:].mean()

307.03623188405794

In [112]:
print(td_allgreen_ai_stats_1['average damage'][7:].mean())
print(td_allred_ai_stats_1['average damage'][14:].mean())
print(td_greenred_ai_stats_1['average damage'][17:].mean())
print(td_redgreen_ai_stats_1['average damage'][20:].mean())

# ---------------------------------------------------------------------------- #

print(td_allgreen_ai_stats_2['average damage'][11:].mean())
#print(td_allred_ai_stats_2['average damage'][7:].mean()
#print(td_greenred_ai_stats_2['average damage'][7:].mean()
#print(td_redgreen_ai_stats_2['average damage'][7:].mean()

# ---------------------------------------------------------------------------- #

print(td_allgreen_ai_stats_3['average damage'][7:].mean())
print(td_allred_ai_stats_3['average damage'][11:].mean())
print(td_greenred_ai_stats_3['average damage'][10:].mean())
print(td_redgreen_ai_stats_3['average damage'][11:].mean())

307.03623188405794
157.78125
463.2564102564104
459.3
478.60526315789474
482.57246376811594
167.43859649122808
288.18333333333334
399.1578947368421


In [ ]:
td_allgreen_rd
td_allred_rd
td_greenred_rd
td_redgreen_rd

ss_yellowmove_rd
ss_yellowstill_rd
ss_redmove_rd
ss_redstill_rd

In [39]:
td_allgreen_rd.iloc[:, -1].mean()

81.42222222222222

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2)
plt.text(x=0.5, y=1.0, s='TD Red+Green Towers - Random vs AI - 30 Runs',
         fontsize=18, ha="center", transform=fig.transFigure)
plt.text(x=0.5, y=0.92, s='Per Run Average and All Run-Single Wave Mean',
         fontsize=12, ha="center", transform=fig.transFigure)
ax1.plot(mean_wave_rg_ai, 'bo', label='AI')
ax1.plot(mean_wave_rg_random, 'ro', label='Random')
ax2.plot(mean_perwave_rg_ai, 'bo', label='AI')
ax2.plot(mean_perwave_rg_random, 'ro', label='Random')
plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)
fig.set_figwidth(15)
plt.show()

print('Média da AI: ' + str(td_redgreen_ai[td_redgreen_ai.columns[-1]].mean()))
print('Desvio Padrão da AI: ' +
      str(td_redgreen_ai[td_redgreen_ai.columns[-1]].std()))
print('Média da Random: ' +
      str(td_redgreen_random[td_redgreen_random.columns[-1]].mean()))
print('Desvio Padrão da Random: ' +
      str(td_redgreen_random[td_redgreen_random.columns[-1]].std()))
